### PR TITLE
[X86] X86FixupVectorConstants - load+sign-extend vector constants that can be stored in a truncated form

### PR DESCRIPTION
**Pull request by RKSimon**
Reduce the size of the vector constant by storing it in the constant pool in a truncated form, and sign-extend it as part of the load.

I've extended the existing FixupConstant functionality to support these sext constant rebuilds - we still select the smallest stored constant entry and prefer vzload/broadcast/vextload for same bitwidth to avoid domain flips.

I intend to add the matching load+zero-extend handling in a future PR, but that requires some alterations to the existing MC shuffle comments handling first.

NOTE: Some of the FixupConstant tables are currently created on the fly as they are dependent on the supported ISAs (HasAVX2 etc.) - should we split these (to allow initializer lists instead) and have duplicate FixupConstant calls to avoid additional stack use?

### CODE REVIEW

File: llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll [hunk 1]
```diff
@@ -750,7 +750,7 @@ define void @vec128_i16_widen_to_i32_factor2_broadcast_to_v4i32_factor4(ptr %in.
 ; AVX512BW-SLOW-LABEL: vec128_i16_widen_to_i32_factor2_broadcast_to_v4i32_factor4:
 ; AVX512BW-SLOW:       # %bb.0:
 ; AVX512BW-SLOW-NEXT:    vmovdqa64 (%rdi), %zmm0
-; AVX512BW-SLOW-NEXT:    vmovdqa {{.*#+}} xmm1 = [0,9,0,11,0,13,0,15]
+; AVX512BW-SLOW-NEXT:    vpmovsxbw {{.*#+}} xmm1 = [0,9,0,11,0,13,0,15]
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: Can small memory size beat defects in micro arch, e.g.

- vpmovsxbw uses one more port
- vmovdqa can be eliminated by uarch

Comment (RKSimon) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: That was my reasoning for always preferring vzload/broadcast over vextload for the same constant load size

Comment (phoebewang) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: I don't get the point. Doesn't vzload have the same TP/ports with vextload?
And if vzload is preferred, why not a `vpmovzxbw` here?
My initial question was about "load size" vs. "inst cost", but we already using broadcast which has the cost too (e.g., vpbroadcastb/w). So it should not be a problem.

Comment (RKSimon) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: This is how I see it regarding preference between vzload/broadcast/vextload:
1. vzload shouldn't ever need a shuffle port to zero the upper elements and the fp/int domain versions are equally available so we don't introduce a domain crossing penalty.
2. broadcast sometimes need a shuffle port (especially for 8/16-bit variants), AVX1 only has fp domain broadcasts but AVX2+ have good fp/int domain equivalents
3. vextload always needs a shuffle port and is only ever int domain

Comment (phoebewang) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: Yes, a comment is great.

> vzload shouldn't ever need a shuffle port to zero the upper elements

But this doesn't match what I got from uops.info, where the `vpmovzx` has the same shuffle port as `vpmovsx`

Comment (RKSimon) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: Do you mean vpmovzx (vzextload which I haven't added yet due to some shuffle comment problems) or vmovd/vmovq (vzload) which just zeros the upper elements?

Comment (RKSimon) [reply llvm/test/CodeGen/X86/any_extend_vector_inreg_of_broadcast.ll:753]: I've added a comment describing the preferences

**Review by phoebewang — APPROVED** [review]
I have checked the code and a dozen test files. No mistakes found so far, so LGTM.
It's better to pre-commit a NFC patch for refactor change related to rebuildZeroUpperCst and rebuildSplatCst to double check no test affect by potential typos.

**Pull request merged**